src/main/java/org/opensearch/knn/index/engine/ResolvedRequiredParameters.java

```diff
@@ -0,0 +1,133 @@
+
+/**
+ * Resolved parameters required for constructing a {@link KNNIndexContext}. If any of these parameters can be null,
+ * then their getters need to be wrapped in an {@link java.util.Optional}
+ */
+public final class ResolvedRequiredParameters {
+    @Getter
+    private final VectorDataType vectorDataType;
+    @Getter
+    private final WorkloadModeConfig mode;
+    @Getter
+    private final SpaceType spaceType;
+    @Getter
+    private final KNNEngine knnEngine;
+    @Getter
+    private final CompressionConfig compressionConfig;
+    @Getter
+    private final Version createdVersion;
+    @Getter
+    private final int dimension;
+    @Nullable
+    private final KNNMethodContext knnMethodContext;
+
+    /**
+     *
+     * @param originalParameters The original user provided parameters
+     * @param settings Settings for the index; passing null will mean that it is not possible to resolve for the legacy
+     * @param createdVersion version that this was created for
+     */
+    public ResolvedRequiredParameters(UserProvidedParameters originalParameters, Settings settings, Version createdVersion) {
+        this.dimension = Objects.requireNonNull(originalParameters.getDimension(), "dimension must be set for ResolvedRequiredParameters");
+        this.vectorDataType = Objects.requireNonNull(
+            originalParameters.getVectorDataType() == null ? VectorDataType.DEFAULT : originalParameters.getVectorDataType(),
+            "vectorDataType must be set for ResolvedRequiredParameters"
+        );
+        this.spaceType = Objects.requireNonNull(
+            SpaceTypeResolver.resolveSpaceType(originalParameters.getKnnMethodContext(), this.vectorDataType),
+            "spaceType must be set for ResolvedRequiredParameters"
+        );
+        this.mode = Objects.requireNonNull(
+            resolveWorkloadModeConfig(originalParameters.getMode()),
+            "mode must be set for ResolvedRequiredParameters"
+        );
+        this.compressionConfig = Objects.requireNonNull(
+            CompressionConfig.fromString(originalParameters.getCompressionLevel()),
+            "compressionConfig must be set for ResolvedRequiredParameters"
+        );
+        boolean isLegacy = computeIsLegacy(originalParameters.getKnnMethodContext(), mode, compressionConfig, vectorDataType, settings);
+        this.knnMethodContext = isLegacy
+            ? createKNNMethodContextFromLegacy(settings, createdVersion)
+            : originalParameters.getKnnMethodContext();
+        this.knnEngine = Objects.requireNonNull(
+            KNNEngineResolver.resolveKNNEngine(knnMethodContext, vectorDataType, mode, compressionConfig),
+            "knnEngine must be set for ResolvedRequiredParameters"
+        );
+        this.createdVersion = Objects.requireNonNull(createdVersion, "createdVersion must be set for ResolvedRequiredParameters");
+    }
+
+    public KNNIndexContext resolveKNNIndexContext(boolean shouldTrain) {
```

Comment:
I think creating resolvedRequiredParameters from UserProvidedParameters is okay.
However, when we crate KNNIndexContext, it is not resolving any value it seem. 
Then, the KNNIndexContext is pass around multiple methods to resolve some of the value inside it. Can it be like providing everything upfront to KNNIndexContext and let KNNIndexContext to figure out how to resolve remaining values? 

Letting all other methods to resolve part of KNNIndexContext makes it hard to understand the resolving logic and flow because it scattered across multiple parts. For example, which value gets resolved first and which value is needed to be resolved first to resolve certain value. Where and when certain value get resolved.